### PR TITLE
NSFS | NC | CLI |bucket delete leaving objects

### PR DESCRIPTION
**Pull request by naveenpaul1**
### Explain the changes
1. S3 bucket delete is possible only for empty buckets.
2. Falied unit test cases after changing the `assert_error` condition to `equal`
3. https://github.com/noobaa/noobaa-core/issues/7699

### Issues: Fixed #xxx / Gap #xxx
1. https://github.com/noobaa/noobaa-core/issues/7699

### Testing Instructions:
1. create a bucket and copy one or more items to it and try to delete.
2. test steps mentioned issue [7699](https://github.com/noobaa/noobaa-core/issues/7699)

- [ ] Doc added/updated
- [X] Tests added

### CODE REVIEW

File: src/cmd/manage_nsfs.js [hunk 1]
```diff
@@ -410,7 +411,7 @@ async function fetch_account_data(argv, from_file) {
         data = JSON.parse(raw_data.toString());
     }
     if (action !== ACTIONS.LIST && action !== ACTIONS.STATUS) _validate_access_keys(argv);
-    if (action === ACTIONS.ADD || action === ACTIONS.STATUS) {
+    if (action === ACTIONS.ADD) {
```
Comment (liranmauda) [inline src/cmd/manage_nsfs.js:414]: removing `ACTIONS.STATUS` will lead to an error when trying to do status based on the access key.
We should keep this

File: src/cmd/manage_nsfs.js [hunk 2]
```diff
@@ -663,6 +664,19 @@ async function get_config_data(config_file_path, show_secrets) {
 ///     VALIDATIONS     ///
 ///////////////////////////
 
+async function is_bucket_empty(fs_context, full_bucket_config_path) {
```
Comment (liranmauda) [inline src/cmd/manage_nsfs.js:667]: Please add the JSDoc comment for the TypeScript

File: src/test/unit_tests/test_nc_nsfs_cli.js [hunk 3]
```diff
@@ -378,10 +396,20 @@ mocha.describe('manage_nsfs cli', function() {
             assert_account(account, account_options);
         });
 
+        mocha.it('cli account status without name and access_key', async function() {
```
Comment (liranmauda) [inline src/test/unit_tests/test_nc_nsfs_cli.js:399]: Should be written using jest as decided a few weeks ago. adding to mocha is not something that we want
@nimrod-becker FYI

**Review by liranmauda — CHANGES_REQUESTED** [review]

File: src/cmd/manage_nsfs.js [hunk 4]
```diff
@@ -663,6 +664,19 @@ async function get_config_data(config_file_path, show_secrets) {
...
+    let data;
+    try {
+        data = await get_config_data(full_bucket_config_path);
+    } catch (err){
+        throw_cli_error(ManageCLIError.NoSuchBucket,  full_bucket_config_path);
+    }
+    const entries = await nb_native().fs.readdir(fs_context, data.path);
```
Comment (liranmauda) [inline src/cmd/manage_nsfs.js:674]: we can add a TODO here, once we can read only a single instance in the directory, and not iterate over the whole directory we should use it.

File: src/sdk/bucketspace_fs.js [hunk 5]
```diff
@@ -388,6 +388,13 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                         name,
                         full_path: path.join(this.fs_root, namespace_bucket_config.write_resource.path) // includes write_resource.path + bucket name (s3 flow)
                     }, object_sdk);
+                } else if (namespace_bucket_config){
+                    // TODO check only for single instance.
```
Comment (romayalon) [inline src/sdk/bucketspace_fs.js:392]: what this TODO means?

File: src/test/unit_tests/test_bucketspace_fs.js [hunk 6]
```diff
@@ -315,6 +316,21 @@ mocha.describe('bucketspace_fs', function() {
...
+        mocha.it('delete_bucket for non empty buckets', async function() {
+            const param = { name: test_not_empty_bucketbucket};
+            await create_bucket(param.name);
+            const bucket_file_path = path.join(new_buckets_path, param.name, 'dummy.txt');
+            await nb_native().fs.writeFile(ACCOUNT_FS_CONFIG, bucket_file_path,
+                Buffer.from(JSON.stringify("data")), {
+                    mode: config.BASE_MODE_FILE,
+                });
+            try {
+                await bucketspace_fs.delete_bucket(param, dummy_object_sdk);
```
Comment (romayalon) [inline src/test/unit_tests/test_bucketspace_fs.js:328]: please add here assert.fail('should have failed with NOT EMPTY')

**Review by romayalon — COMMENTED** [review]
looks good, final comments

File: src/sdk/bucketspace_fs.js (hunk 5 above)
Comment (naveenpaul1) [reply src/sdk/bucketspace_fs.js:392]: @romayalon bucket delete should not happen even if there is a single file in bucket, in that case we dont need to iterate through complete files, It was a suggestion from Liran. Will update the comment with more explaining words.

Comment (romayalon) [reply src/sdk/bucketspace_fs.js:392]: that reminds me of something that we need to handle here, if the directory only contains the internal hidden directory we should allow the deletion
few of our internal dirs are - 
get_bucket_tmpdir(), versions directory

Comment (liranmauda) [reply src/sdk/bucketspace_fs.js:392]: I asked for the TODO, when we will implement the object list optimization we can call readdir whiteout iterating over all the files. 
Then we should implement it also here, as this readdir is a performance hazard... 
If we have a lot of files/objects, then failing this delete bucket for the reason that the bucket is not empty will take a lot of time.

Comment (romayalon) [reply src/sdk/bucketspace_fs.js:392]: currently on Namespace FS we do like follows - 
```
  const list = await this.list_objects({ ...params, limit: 1 }, object_sdk);

  if (list && list.objects && list.objects.length > 0) {
      throw new RpcError('NOT_EMPTY', 'underlying directory has files in it');
  }
```
I guess that what @liranmauda means, list_objects ignores also the hidden directories @naveenpaul1 you can do this here as well, just replace this with the correct namespace

Comment (liranmauda) [reply src/sdk/bucketspace_fs.js:392]: @romayalon This is a great sulotion

Comment (naveenpaul1) [reply src/sdk/bucketspace_fs.js]: @romayalon @liranmauda updated code with `ns.list_object()` to get the dir items

**Review by romayalon — APPROVED** [review]

**Review by liranmauda — APPROVED** [review]

**Pull request merged**